File: android/src/main/kotlin/com/jiguang/joperate_flutter_plugin/JoperateFlutterPlugin.kt
package com.jiguang.joperate_flutter_plugin

import android.content.Context
import androidx.annotation.NonNull
import cn.jiguang.api.utils.JCollectionAuth
import cn.jiguang.joperate.api.JOperateInterface

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

/** JoperateFlutterPlugin */
class JoperateFlutterPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private var context: Context? = null


  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "joperate_flutter_plugin")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext

  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    }else if (call.method == "init") {
      init(call)
    }else if (call.method == "eventRecord") {
      eventRecord(call)
    }else if (call.method == "identify") {
      identifyAccount(call, result)
    }else if (call.method == "setUserChannel") {
      setUserChannel(call, result)
    }else if (call.method == "CUID") {
      CUID(call, result)
    }else if (call.method == "setUtmProperties") {
      setUtmProperties(call, result)
    }else if (call.method == "setCommonProperties") {
      setCommonProperties(call, result)
    }else if (call.method == "setDynamicCommonProperties") {
      setDynamicCommonProperties(call, result)
    }else if (call.method == "unregisterCommonProperty") {
      unregisterCommonProperty(call, result)
    }else if (call.method == "clearCommonProperties") {
      clearCommonProperties(call, result)
    }else if (call.method == "set") {
      set(call, result)
    }else if (call.method == "setOnce") {
      setOnce(call, result)
    }else if (call.method == "increment") {
      increment(call, result)
    }else if (call.method == "append") {
      append(call, result)
    }else if (call.method == "unset") {
      unset(call, result)
    }else if (call.method == "timeKeepingEventStart") {
      timeKeepingEventStart(call, result)
    }else if (call.method == "timeKeepingEventEnd") {
      timeKeepingEventEnd(call, result)
    }else if (call.method == "timeKeepingEventPause") {
      timeKeepingEventPause(call, result)
    }else if (call.method == "timeKeepingEventResume") {
      timeKeepingEventResume(call, result)
    }else if (call.method == "removeTimeKeepingEvent") {
      removeTimeKeepingEvent(call, result)
    }else if (call.method == "clearTimeKeepingEvent") {
      clearTimeKeepingEvent(call, result)
    }else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  fun init(call: MethodCall){
    val arguments = call.arguments as HashMap<*, *>
    if (arguments["debug"] == true) {
      JOperateInterface.setDebug(true)
    }
    JCollectionAuth.setAuth(context, true)
    JOperateInterface.getInstance(context).init()
  }

  fun eventRecord(call: MethodCall){
    val arguments = call.arguments as HashMap<*, *>
    var json: JSONObject? = null
    if (arguments["property"] != null) {
      json = JSONObject(arguments["property"] as HashMap<*, *>)
    }
    JOperateInterface.getInstance(context).onEvent(arguments["eventName"] as String?, json)
  }

  fun identifyAccount(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    if (json.length() > 0) {
      JOperateInterface.getInstance(context).login(json) { i, s ->
        val map = HashMap<Any, Any?>()
        map["errcode"] = i;
        map["msg"] = s;
        result.success(map)
      }
    }
  }

  fun setUserChannel(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    if (json.length() > 0) {
      JOperateInterface.getInstance(context).setChannel(json) { i, s ->
        val map = HashMap<Any, Any?>()
        map["errcode"] = i;
        map["msg"] = s;
        result.success(map)
      }
    }
  }

  fun CUID(call: MethodCall, result: Result) {
    result.success(JOperateInterface.getInstance(context).cuid)
  }

  fun setUtmProperties(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    if (json.length() > 0) {
      JOperateInterface.getInstance(context).setUtmProperties(json)
    }
  }

  fun setCommonProperties(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    if (json.length() > 0) {
      JOperateInterface.getInstance(context).registerSuperProperties(json)
    }
  }

  fun setDynamicCommonProperties(call: MethodCall, result: Result) {
    val f = call.arguments as () -> JSONObject
    JOperateInterface.getInstance(context).registerDynamicSuperProperties {
      return@registerDynamicSuperProperties f()
    }
  }

  fun unregisterCommonProperty(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).unregisterSuperProperty(call.arguments as String?)
  }

  fun clearCommonProperties(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).clearSuperProperties()
  }

  fun set(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    JOperateInterface.getInstance(context).profileSet(json){ i, s ->
      val map = HashMap<Any, Any>()
      map["errcode"] = i;
      map["msg"] = s;
      result.success(map)
    }
  }

  fun setOnce(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val json = JSONObject(arguments)
    JOperateInterface.getInstance(context).profileSetOnce(json){ i, s ->
      val map = HashMap<Any, Any>()
      map["errcode"] = i;
      map["msg"] = s;
      result.success(map)
    }
  }

  fun increment(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<String, Number>
    JOperateInterface.getInstance(context).profileIncrement(arguments){ i, s ->
      val map = HashMap<Any, Any>()
      map["errcode"] = i;
      map["msg"] = s;
      result.success(map)
    }
  }

  fun append(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    val list = arguments["content"] as List<String>

    JOperateInterface.getInstance(context).profileAppend(arguments["key"] as String, list.toSet()){ i, s ->
      val map = HashMap<Any, Any>()
      map["errcode"] = i;
      map["msg"] = s;
      result.success(map)
    }
  }

  fun unset(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).profileUnset(call.arguments as String){ i, s ->
      val map = HashMap<Any, Any>()
      map["errcode"] = i;
      map["msg"] = s;
      result.success(map)
    }
  }

  fun timeKeepingEventStart(call: MethodCall, result: Result) {
    val key = JOperateInterface.getInstance(context).onEventTimerStart(call.arguments as String)
    result.success(key)
  }

  fun timeKeepingEventEnd(call: MethodCall, result: Result) {
    val arguments = call.arguments as HashMap<*, *>
    var json: JSONObject? = null
    if (arguments["properties"] != null) {
      json = JSONObject(arguments["properties"] as HashMap<*, *>)
    }
    JOperateInterface.getInstance(context).onEventTimerEnd(arguments["eventId"] as String, json)
  }

  fun timeKeepingEventPause(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).onEventTimerPause(call.arguments as String)
  }

  fun timeKeepingEventResume(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).onEventTimerResume(call.arguments as String)
  }

  fun removeTimeKeepingEvent(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).removeTimer(call.arguments as String)
  }

  fun clearTimeKeepingEvent(call: MethodCall, result: Result) {
    JOperateInterface.getInstance(context).clearTrackTimer()
  }
}

File: ios/Classes/JoperateFlutterPlugin.m
#import "JoperateFlutterPlugin.h"
#import "JOPERATEService.h"

@implementation JoperateFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"joperate_flutter_plugin"
            binaryMessenger:[registrar messenger]];
  JoperateFlutterPlugin* instance = [[JoperateFlutterPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"init" isEqualToString:call.method]) {
      [self setupWithConfig:call];
  } else if ([@"eventRecord" isEqualToString:call.method]) {
      [self eventRecord:call];
  } else if ([@"identify" isEqualToString:call.method]) {
      [self identifyAccount:call result:result];
  } else if ([@"setUserChannel" isEqualToString:call.method]) {
      [self setUserChannel:call result:result];
  } else if ([@"CUID" isEqualToString:call.method]) {
      [self CUID:result];
  } else if ([@"setUtmProperties" isEqualToString:call.method]) {
      [self setUtmProperties:call];
  } else if ([@"setCommonProperties" isEqualToString:call.method]) {
      [self setCommonProperties:call];
  } else if ([@"setDynamicCommonProperties" isEqualToString:call.method]) {
      [self setDynamicCommonProperties:call];
  } else if ([@"unregisterCommonProperty" isEqualToString:call.method]) {
      [self unregisterCommonProperty:call];
  } else if ([@"clearCommonProperties" isEqualToString:call.method]) {
      [self clearCommonProperties];
  } else if ([@"currentCommonProperties" isEqualToString:call.method]) {
      [self currentCommonProperties:result];
  } else if ([@"set" isEqualToString:call.method]) {
      [self set:call result:result];
  } else if ([@"setOnce" isEqualToString:call.method]) {
      [self setOnce:call result:result];
  } else if ([@"increment" isEqualToString:call.method]) {
      [self increment:call result:result];
  } else if ([@"append" isEqualToString:call.method]) {
      [self append:call result:result];
  } else if ([@"unset" isEqualToString:call.method]) {
      [self unset:call result:result];
  } else if ([@"timeKeepingEventStart" isEqualToString:call.method]) {
      [self timeKeepingEventStart:call result:result];
  } else if ([@"timeKeepingEventEnd" isEqualToString:call.method]) {
      [self timeKeepingEventEnd:call result:result];
  } else if ([@"timeKeepingEventPause" isEqualToString:call.method]) {
      [self timeKeepingEventPause:call];
  } else if ([@"timeKeepingEventResume" isEqualToString:call.method]) {
      [self timeKeepingEventResume:call];
  } else if ([@"removeTimeKeepingEvent" isEqualToString:call.method]) {
      [self removeTimeKeepingEvent:call];
  } else if ([@"clearTimeKeepingEvent" isEqualToString:call.method]) {
      [self clearTimeKeepingEvent];
  } else {
    result(FlutterMethodNotImplemented);
  }
}


- (void)setupWithConfig:(FlutterMethodCall*)call {
    NSDictionary * arguments = call.arguments;
    NSNumber *debug = arguments[@"debug"];
    if ([debug boolValue]) {
        [JOPERATEService setDebug:YES];
    } else {
        [JOPERATEService setDebug:NO];
    }
    JOPERATEConfig * config = [[JOPERATEConfig alloc] init];
    config.appKey = [arguments objectForKey:@"appkey"];
    config.channel = [arguments objectForKey:@"channel"];
    [JOPERATEService setupWithConfig:config];
}

- (void)eventRecord:(FlutterMethodCall*)call {
    NSDictionary * arguments = call.arguments;
    JOPERATEEventObject * event = [[JOPERATEEventObject alloc] init];
    event.eventName = arguments[@"eventName"];
    event.property = arguments[@"property"];
    [JOPERATEService eventRecord:event];
}

- (void)identifyAccount:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    JOPERATEUserID * userID = [[JOPERATEUserID alloc] init];
    [userID setCompletion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
    userID.userIDs = arguments;
    [JOPERATEService identifyAccount:userID];
}

- (void)setUserChannel:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    JOPERATEUserChannel * channel = [[JOPERATEUserChannel alloc] init];
    [channel setCompletion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
    channel.channels = arguments;
    [JOPERATEService setUserChannel:channel];
}

- (void)CUID:(FlutterResult)result {
    NSLog(@"%@", ([JOPERATEService CUID]));
    result([JOPERATEService CUID]);
}

- (void)setUtmProperties:(FlutterMethodCall*)call {
    NSDictionary * arguments = call.arguments;
    [JOPERATEService setUtmProperties:arguments];
}

- (void)setCommonProperties:(FlutterMethodCall*)call {
    NSDictionary * arguments = call.arguments;
    [JOPERATEService setCommonProperties:arguments];
}
 
- (void)setDynamicCommonProperties:(FlutterMethodCall*)call {
    [JOPERATEService setDynamicCommonProperties:call.arguments];
}

- (void)unregisterCommonProperty:(FlutterMethodCall*)call {
    [JOPERATEService unregisterCommonProperty:call.arguments];
}
 
- (void)clearCommonProperties {
    [JOPERATEService clearCommonProperties];
}

- (void)currentCommonProperties:(FlutterResult)result {
    result([JOPERATEService currentCommonProperties]);
}

- (void)set:(FlutterMethodCall*)call result:(FlutterResult)result {
    [JOPERATEService set:call.arguments completion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
}

- (void)setOnce:(FlutterMethodCall*)call result:(FlutterResult)result {
    [JOPERATEService setOnce:call.arguments completion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
}
 
- (void)increment:(FlutterMethodCall*)call result:(FlutterResult)result {
    [JOPERATEService increment:call.arguments completion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
}
 
- (void)append:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    [JOPERATEService append:arguments[@"key"] by:arguments[@"content"] completion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
}
     
- (void)unset:(FlutterMethodCall*)call result:(FlutterResult)result {
    [JOPERATEService unset:call.arguments completion:^(NSInteger errcode, NSString * _Nonnull msg) {
        result(@{@"errcode": @(errcode), @"msg": msg ?: @""});
    }];
}
 
- (void)timeKeepingEventStart:(FlutterMethodCall*)call result:(FlutterResult)result {
    result([JOPERATEService timeKeepingEventStart:call.arguments]);
}
 
- (void)timeKeepingEventEnd:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    [JOPERATEService timeKeepingEventEnd:arguments[@"eventId"] with:arguments[@"properties"]];
}

- (void)timeKeepingEventPause:(FlutterMethodCall*)call {
    [JOPERATEService timeKeepingEventPause:call.arguments];
}
 
- (void)timeKeepingEventResume:(FlutterMethodCall*)call {
    [JOPERATEService timeKeepingEventResume:call.arguments];
}
 
- (void)removeTimeKeepingEvent:(FlutterMethodCall*)call {
    [JOPERATEService removeTimeKeepingEvent:call.arguments];
}
 
- (void)clearTimeKeepingEvent {
    [JOPERATEService clearTimeKeepingEvent];
}

@end
